### PR TITLE
Oceanwater 794 health monitor

### PR DESCRIPTION
**Pull request by Keeganfn**
## Summary

Health monitor PLEXIL plan that uses the fault dependencies feature to make it easier for other plans to monitor faults. 

This PR also includes 2 new PLEXIL commands: ```inject_simulated_fault``` and ```clear_simulated_fault```. This allows the user to inject or clear any supported non-goal fault from a PLEXIL plan. It is intended for use in testing and simulation, and should not be used to actually clear a hardware fault in a real plan. They also have the ability to make triggering and clearing a fault stochastic (based on a given probability), so that plans can simulate a degree of randomness.

 ### Commands:

-  ```inject_simulated_fault``` :  This command takes two arguments: ```fault_name``` (string) and ```probability``` (Real) which is the probability of the fault being succesfully injected (range 0-1).
-  ```clear_simulated_fault``` :  This command takes two arguments: ```fault_name``` (string) and ```probability``` (Real) which is the probability of the fault being succesfully cleared (range 0-1).

Both commands return a Boolean which represents if the action was succesfully completed. Returning True means that the probability check passed and the fault was succesfully injected or cleared. Returning False means that the probability check failed and simulates being unable to inject or clear the fault. 

To not have any randomness the probability should be set to 1.

## Build
1. ow_simulator on master branch
2. ```catkin clean```
3. ```catkin build```

## Testing
There are 3 new PLEXIL plans: ```TestSimulatedFaults```, ```HealthMonitor``` and ```HealthMonitorDemo```. 

**Test1 (Simulated Faults):** 
1. ```roslaunch ow_plexil ow_exec.launch plan:="TestSimulatedFaults.plx" fault_dependencies_file:="FaultDependenciesModel.xml"```
2. This plan runs 3 tests. Test 1 injects and clears antenna pan fault with 100% probability. Test 2 injects an antenna pan fault, and repeatedly tries to clear it with 25% probability until the fault is succesfully cleared. Test 3 is the same as test 2 but the injection command is repeated until successful instead of the clear command.
3. You should see messages every time a inject or clear command is called, as well as a fault warning message if a fault was succesfully injected or cleared.

Note: I ran into the SynchronousCommand bug again with these commands, so they are asynchronous command calls with a 1 second wait after each one. 

**Test2 (HealthMonitorDemo):** 
1. ```roslaunch ow_plexil ow_exec.launch plan:="HealthMonitorDemo.plx" fault_dependencies_file:="FaultDependenciesModel.xml"```
2.  Demo unstows and stows the arm if the Camera is set as operable by the HealthMonitor. If the Arm is inoperable, and the Camera is operable it takes photos repeatedly. If both the Camera and Arm are inoperable, it prints out an error message and waits for them to be operable.
3. Injecting an Arm fault will trigger the picture action, if you then trigger a Camera fault as well it will print out the error message. Any combination of fault injection should work. Every time a subsystem becomes Operable or Inoperable, the HealthMonitor will print out the impacted subsystem status and the cause. 
4. Cause a Low State of Charge fault, this should make every subsystem inoperable. All subsystems other than power should state that it is due to a non-local fault, and the Power subsystem should print out that the cause is a local fault + the fault name. 

Let me know if you have any questions/comments! Thank you!

### CODE REVIEW

**Comment by kmdalal** [comment]
@Keeganfn, please clarify "Both commands return a Boolean which represents if the action was succesful or not." in the instructions above?  Not sure if 'successful' (note typo in original) means that the fault was set/cleared.  Also, the second command should probably be 'clear...`.

**Comment by kmdalal** [comment]
I think I found a bug, which probably existed in the last branch.  When I run the HealthMonitorDemo plan, I always get `[ WARN]: Could not use IsOperable lookup, fault dependencies are turned off`.  I definitely loaded the dependencies, because they worked on the previous test.  But I'm loading plans using the GUI, and not on the command line with a new roslaunch...

**Comment by Keeganfn** [comment]
> I think I found a bug, which probably existed in the last branch. When I run the HealthMonitorDemo plan, I always get `[ WARN]: Could not use IsOperable lookup, fault dependencies are turned off`. I definitely loaded the dependencies, because they worked on the previous test. But I'm loading plans using the GUI, and not on the command line with a new roslaunch...

Hm that is odd, I am unable to recreate this error, can I see the roslaunch command you are using as well as the output?

**Comment by kmdalal** [comment]
The previous reported problem actually occurs when the test is performed as written.  My branch is clean and rebuilt...
```
$ roslaunch ow_plexil ow_exec.launch plan:="HealthMonitorDemo.plx" fault_dependencies:="FaultDependenciesModel.xml"
... logging to /home/kdalal/.ros/log/51443dc4-45d8-11ee-b132-c3af97c1d924/roslaunch-rockaway.ndc.nasa.gov-148915.log
Checking log directory for disk usage. This may take a while.
Press Ctrl-C to interrupt
Done checking log file disk usage. Usage is <1GB.

started roslaunch server http://rockaway.ndc.nasa.gov:46665/

SUMMARY
========

PARAMETERS
 * /fault_dependencies_file: 
 * /fault_dependencies_path: /usr/local/home/k...
 * /fault_dependencies_verbose: False
 * /owlat_flag: False
 * /rosdistro: noetic
 * /rosversion: 1.16.0

NODES
  /
    identify_sample_location (ow_plexil/identify_sample_location.py)
    ow_exec_node (ow_plexil/ow_exec_node)
    terminal_selection_node (ow_plexil/terminal_selection_node)

ROS_MASTER_URI=http://localhost:11311

process[ow_exec_node-1]: started with pid [148940]
process[terminal_selection_node-2]: started with pid [148941]
process[identify_sample_location-3]: started with pid [148942]
[ INFO]: Starting PLEXIL executive node...
[ INFO]: Executive node started, ready for PLEXIL plans.
[ INFO]: PLEXIL: Starting HealthMonitorDemo...
[ WARN]: Could not use IsOperable lookup, fault dependencies are turned off
[ INFO]: PLEXIL: Unstowing and stowing arm...
[ INFO]: ArmUnstow () started...
[ INFO]: ArmUnstow finished in state SUCCEEDED
[ INFO]: ArmStow () started...
^C[identify_sample_location-3] killing on exit
[terminal_selection_node-2] killing on exit
[ow_exec_node-1] killing on exit
[ow_exec_node-1] escalating to SIGTERM
shutting down processing monitor...
... shutting down processing monitor complete
```

**Comment by Keeganfn** [comment]
> The previous reported problem actually occurs when the test is performed as written. My branch is clean and rebuilt...
> 
> ```
> $ roslaunch ow_plexil ow_exec.launch plan:="HealthMonitorDemo.plx" fault_dependencies:="FaultDependenciesModel.xml"
> ... logging to /home/kdalal/.ros/log/51443dc4-45d8-11ee-b132-c3af97c1d924/roslaunch-rockaway.ndc.nasa.gov-148915.log
> Checking log directory for disk usage. This may take a while.
> Press Ctrl-C to interrupt
> Done checking log file disk usage. Usage is <1GB.
> 
> started roslaunch server http://rockaway.ndc.nasa.gov:46665/
> 
> SUMMARY
> ========
> 
> PARAMETERS
>  * /fault_dependencies_file: 
>  * /fault_dependencies_path: /usr/local/home/k...
>  * /fault_dependencies_verbose: False
>  * /owlat_flag: False
>  * /rosdistro: noetic
>  * /rosversion: 1.16.0
> 
> NODES
>   /
>     identify_sample_location (ow_plexil/identify_sample_location.py)
>     ow_exec_node (ow_plexil/ow_exec_node)
>     terminal_selection_node (ow_plexil/terminal_selection_node)
> 
> ROS_MASTER_URI=http://localhost:11311
> 
> process[ow_exec_node-1]: started with pid [148940]
> process[terminal_selection_node-2]: started with pid [148941]
> process[identify_sample_location-3]: started with pid [148942]
> [ INFO]: Starting PLEXIL executive node...
> [ INFO]: Executive node started, ready for PLEXIL plans.
> [ INFO]: PLEXIL: Starting HealthMonitorDemo...
> [ WARN]: Could not use IsOperable lookup, fault dependencies are turned off
> [ INFO]: PLEXIL: Unstowing and stowing arm...
> [ INFO]: ArmUnstow () started...
> [ INFO]: ArmUnstow finished in state SUCCEEDED
> [ INFO]: ArmStow () started...
> ^C[identify_sample_location-3] killing on exit
> [terminal_selection_node-2] killing on exit
> [ow_exec_node-1] killing on exit
> [ow_exec_node-1] escalating to SIGTERM
> shutting down processing monitor...
> ... shutting down processing monitor complete
> ```

Ah I think I see the issue, I incorrectly used the old argument name in the test instructions: ```fault_dependencies``` it was changed to: ```fault_dependencies_file```. 

I have updated the instructions, sorry about that!

**Comment by kmdalal** [comment]
Test 2 now works as instructed.  But if I load the `HealthMonitorDemo` plan after having run Test 1 in the same roslaunch, I get the same warning "Could not use IsOperable lookup, fault dependencies are turned off" repeatedly...

File: ow_plexil/src/plans/HealthMonitor.plp [hunk 1]
```diff
@@ -0,0 +1,98 @@
...
+      }
+      else{
+        log_info("Arm is Inoperable due to non-local fault(s).");
+      }
+      PrevArm = false;
+    }
+    if ArmOperable && !PrevArm{
+        log_info("Arm is now operable.");
+        PrevArm = true;
+    }
```
Comment (kmdalal) [inline ow_plexil/src/plans/HealthMonitor.plp:53]: I think the four similar instances of this code block could be replaced with a Library node of 3 parameters.  Optional since you are pressed for time!

File: ow_plexil/src/plans/HealthMonitorDemo.plp [hunk 2]
```diff
@@ -0,0 +1,61 @@
...
+// error message and waits for them to be operable. 
+
+#include "plan-interface.h"
+
+LibraryAction HealthMonitor (InOut Boolean AllOperable,
+                             InOut Boolean ArmOperable,
+                             InOut Boolean AntennaOperable,
+                             InOut Boolean CameraOperable,
+                             InOut Boolean PowerOperable);
+
```
Comment (kmdalal) [inline ow_plexil/src/plans/HealthMonitorDemo.plp:17]: I suggest adding this declaration to `plan-interface.h`.

File: ow_plexil/src/plans/HealthMonitorDemo.plp [hunk 3]
```diff
@@ -0,0 +1,61 @@
...
+  
+  log_info("Starting HealthMonitorDemo...");
+
+  Monitor:
+  {
+  LibraryCall HealthMonitor (AllOperable = AllGood,
+                             ArmOperable = ArmGood,
+                             AntennaOperable = AntennaGood,
+                             CameraOperable = CameraGood,
+                             PowerOperable = PowerGood);
```
Comment (kmdalal) [inline ow_plexil/src/plans/HealthMonitorDemo.plp:35]: Minor: this block should be indented.

File: ow_plexil/src/plans/HealthMonitorDemo.plp [hunk 4]
```diff
@@ -0,0 +1,61 @@
...
+  }
+
+  DemoPattern:
+  {
+    Repeat true;
+    Wait 3;
```
Comment (kmdalal) [inline ow_plexil/src/plans/HealthMonitorDemo.plp:41]: It would be useful to parameterize the Wait duration, however I'm not sure if Wait accepts an integer variable.  And in any case, not needed for this PR.

File: ow_plexil/src/plans/HealthMonitorDemo.plp [hunk 5]
```diff
@@ -0,0 +1,61 @@
...
+
+    if ArmGood{
+      Invariant ArmGood;
+      log_info("Unstowing and stowing arm...");
+      LibraryCall ArmUnstow();
+      LibraryCall ArmStow();
```
Comment (kmdalal) [inline ow_plexil/src/plans/HealthMonitorDemo.plp:47]: I wonder if this block is susceptible to the "SynchronousCommand problem", but that's out of scope of this PR.

File: ow_plexil/src/plans/HealthMonitorDemo.plp [hunk 6]
```diff
@@ -0,0 +1,61 @@
...
+    }
+    else if CameraGood{
+      log_info("Arm not operable, attempting CameraCapture to diagnose...");
+      LibraryCall CameraCapture();
```
Comment (kmdalal) [inline ow_plexil/src/plans/HealthMonitorDemo.plp:51]: It would be really cool if the camera tracked the arm (end effector) as it moved, and I think @AstroStucky wrote such a plan, but this is out of scope of this PR and the current behavior is fine.

File: ow_plexil/src/plexil-adapter/OwInterface.cpp [hunk 7]
```diff
@@ -908,6 +908,41 @@ void OwInterface::taskDiscardSampleAction (int frame, bool relative,
      default_action_done_cb<TaskDiscardSampleResultConstPtr> (Name_TaskDiscardSample));
 }
 
+bool OwInterface::injectSimulatedFault (const std::string &fault_name, double probability) const
+{
+  srand(time(0));
+  double roll = (double) rand()/RAND_MAX;
+  string combined_fault_name = "faults/" + fault_name;
+  if (roll <= probability){
+    if(m_genericNodeHandle->hasParam(combined_fault_name)){
```
Comment (kmdalal) [inline ow_plexil/src/plexil-adapter/OwInterface.cpp:917]: I think these two checks should be reversed, so that the function returns false if the fault name is invalid, regardless of the roll.  The roll should only determine if the fault is set.  

This comment also applies to the next function.

File: ow_plexil/src/plans/HealthMonitorDemo.plp (hunk 5 above)
Comment (Keeganfn) [reply ow_plexil/src/plans/HealthMonitorDemo.plp:41]: I would as well, interestingly enough, I havent ran into the SynchronousCommand problem yet for this plan. Its strange because normally the ArmUnstow() and ArmStow() always have that issue.

**Review by kmdalal — APPROVED** [review]
Tests work well (the given ones plus a few of my own), code looks good.  Great work!
